src/audio/bgm_control.c

```diff
@@ -0,0 +1,439 @@
+                            music->state = stateFadeOut;
+                        }
+                    }
+                } else {
+                    if (music->flags & MUSIC_FLAG_PUSHING) {
+                        // nothing was playing in this case, so remember to skip the next pop
+                        music->flags |= MUSIC_FLAG_IGNORE_POP;
+                    }
+                    music->flags &= ~pushedFlag;
+                    music->state = MUSIC_STATE_PLAY_NEXT;
+                }
+                break;
+            case MUSIC_STATE_AWAIT_FADEOUT:
+                flags = music->flags;
+                music->flags &= ~pushedFlag;
+                if (flags & MUSIC_FLAG_PLAYING) {
+                    if (snd_song_is_playing(music->songName) == AU_RESULT_OK) {
+                        music->flags &= ~MUSIC_FLAG_PLAYING;
+                        music->state = MUSIC_STATE_DELAY_2;
+                    }
+                } else {
+                    music->state = MUSIC_STATE_PLAY_NEXT;
+                }
+                break;
+            case MUSIC_STATE_DELAY_2:
+                music->state = MUSIC_STATE_DELAY_1;
+                break;
+            case MUSIC_STATE_DELAY_1:
+                music->state = MUSIC_STATE_PLAY_NEXT;
+                break;
+            case MUSIC_STATE_PLAY_NEXT:
+                if (!(music->flags & MUSIC_FLAG_POPPING)) {
+                    if (music->requestedSongID <= AU_SONG_NONE) {
+                        // new song is AU_SONG_NONE, play nothing
+                        music->state = MUSIC_STATE_IDLE;
+                    } else {
+                        music->songName = snd_song_load(music->requestedSongID, i);
+                        // snd_song_lond may return either songName or an AuResult error code.
+                        // since song names are 4-character big-endian identifiers packed into an s32, we require
+                        // the upper half of the return value to be nonzero for songs which loaded without error.
+                        // this reserves return values from 0 to 0xFFFF for AuResult codes.
+                        if (music->songName > 0xFFFFU) {
+                            if ((music->flags & MUSIC_FLAG_FADE_IN_NEXT)) {
+                                snd_song_request_fade_in(music->songName, music->variation,
+                                    music->fadeInTime, music->fadeStartVolume, music->fadeEndVolume);
+                                music->flags &= ~MUSIC_FLAG_FADE_IN_NEXT;
+                            } else {
+                                // if we aren't fading in, just start at the default volume level
+                                bgm_set_target_volume(MusicDefaultVolume);
+                            }
+                            /// @bug this is called even if we are trying to fade in, immediately overriding any fade parameters
```

Comment:
Just as a note, this is an interesting find here. This is why the script func FadeInMusic doesn't work (not sure whether this is known, even by the developers who call it from scripts as though it will work). To fix, this call just needs to be moved into the preceding else case.